File: Libraries/Megamark Unity/Assets/Megamark Controller/Demo Scripts/RobotExplorerTest.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

/*
 * This example demonstrates coordination between sensor data and actuator data.
 * The robot will move forward until it detects an obstacle in front of it with the laser.
 * While the obstacle is detected, the robot will rotate right until the obstacle is no longer detected. 
 * This example assumes there is a laser on pin A4 attached on the right gripper facing forward. 
 */

/*** NOTE: THIS EXAMPLE DOES NOT WORK, AS UNITY'S SERIAL.READLINE COMMAND FREEZES UNITY! ***/

public class RobotExplorerTest : MonoBehaviour {

	public Megamark megamark;
	private IEnumerator coroutine;

	public bool obstacleDetected = true;

	// Use this for initialization
	void Start () {
		coroutine = Explore();
		StartCoroutine(coroutine);
	}

	// Check whether an obstacle has been detected. If so, rotate to avoid obstacle.
	IEnumerator Explore() {
		while (true) {
			
			int right = megamark.getRightLaser();

			// move forward if no obstacle detected
			if (right > 200) {
				megamark.setWheelVelocities(0.2f,0.2f);
				if(obstacleDetected == true) {
					obstacleDetected = false;
					Debug.Log("No obstacle detected! Moving forward...");
				}
			}
			// rotate right if obstacle detected
			else {
				megamark.setWheelVelocities(0.2f,-0.2f);
				if(obstacleDetected == false) {
					obstacleDetected = true;
					Debug.Log("Obstacle detected at " + right + "mm! Rotating right...");
				}
			}

		}
	}
		
}

File: Libraries/Megamark RobotC/RobotExplorerTest.c
#pragma config(CircuitBoardType, typeCktBoardMega)
#pragma config(Motor,  servo_3,         rightShoulder, tmotorServoStandard, openLoop, IOPins, dgtl22, None)
#pragma config(Motor,  servo_6,         rightElbow,    tmotorServoStandard, openLoop, IOPins, dgtl24, None)
#pragma config(Motor,  servo_13,        R1,            tmotorServoStandard, openLoop, IOPins, dgtl26, None)
#pragma config(Motor,  motor_4,         R2,            tmotorServoStandard, openLoop, IOPins, dgtl28, None)
#pragma config(Motor,  motor_9,         R3,            tmotorServoStandard, openLoop, IOPins, dgtl30, None)
#pragma config(Motor,  motor_10,        R4,            tmotorServoStandard, openLoop, IOPins, dgtl32, None)
#pragma config(Sensor, dgtl5,  leftDir,           sensorDigitalOut)
#pragma config(Sensor, dgtl4,  leftOn,           sensorDigitalOut)
#pragma config(Sensor, dgtl3,  rightDir,           sensorDigitalOut)
#pragma config(Sensor, dgtl2,  rightOn,           sensorDigitalOut)
#pragma config(Sensor, anlg4,  rightLaser,     sensorAnalog)

/* This example demonstrates coordination between sensor data and actuator data.
 * The robot will move forward until it detects an obstacle in front of it with the laser.
 * While the obstacle is detected, the robot will rotate right until the obstacle is no longer detected.
 * This example assumes there is a laser on pin A4 attached on the right gripper facing forward. */

task main()
{
  while(true)
  {
    //move forward if no obstacle is detected
    if(SensorValue[rightLaser] > 200)
    {
      SensorValue[leftDir]  = true;  //Left wheel forward direction.
      SensorValue[leftOn]   = true;  //Left wheel full speed.
      SensorValue[rightDir] = true;  //Right wheel forward direction.
      SensorValue[rightOn]  = true;  //Right wheel full speed.
    }
    //rotate right if obstacle is detected
    else
    {
      SensorValue[leftDir]  = true;  //Left wheel forward direction.
      SensorValue[leftOn]   = true;  //Left wheel full speed.
      SensorValue[rightDir] = false; //Right wheel backward direction.
      SensorValue[rightOn]  = true;  //Right wheel full speed.
    }
  }
}
